projects/packages/account-protection/src/class-validation-service.php

```diff
@@ -49,6 +49,149 @@ protected function request_suffixes( string $password_prefix ) {
+			$errors[] = __( 'Doesn\'t match user data', 'jetpack-account-protection' );
+		}
+
+		if ( $this->is_recent_password( $user->ID, $password ) ) {
+			$errors[] = __( 'Not used recently', 'jetpack-account-protection' );
+		}
+
+		if ( $this->is_weak_password( $password ) ) {
+			$errors[] = __( 'Not a leaked password.', 'jetpack-account-protection' );
+		}
+
+		return $errors;
+	}
+
+	/**
+	 * Return first validation error.
+	 *
+	 * @param \WP_User|\stdClass $user The user object or a copy.
+	 * @param string             $password The password to check.
+	 * @param 'profile'|'reset'  $context The context the validation is run in.
+	 *
+	 * @return string The first validation errors (if any).
+	 */
+	public function return_first_validation_error( $user, string $password, $context ): string {
+		if ( 'profile' === $context ) {
+			if ( empty( $password ) ) {
+				return __( '<strong>Error:</strong> The password cannot be a space or all spaces.', 'jetpack-account-protection' );
+			}
+		}
+
+		if ( 'reset' === $context ) {
+			if ( $this->contains_backslash( $password ) ) {
+				return __( '<strong>Error:</strong> The password cannot contain a backslash (\\) character.', 'jetpack-account-protection' );
+			}
+		}
+
+		if ( $this->is_invalid_length( $password ) ) {
+			return __( '<strong>Error:</strong> The password must be between 6 and 150 characters.', 'jetpack-account-protection' );
+		}
+
+		if ( $this->matches_user_data( $user, $password ) ) {
+			return __( '<strong>Error:</strong> The password matches user data.', 'jetpack-account-protection' );
+		}
+
+		if ( $this->is_recent_password( $user->ID, $password ) ) {
+			return __( '<strong>Error:</strong> The password was used recently.', 'jetpack-account-protection' );
+		}
+
+		if ( $this->is_weak_password( $password ) ) {
+			return __( '<strong>Error:</strong> The password was found in a public leak.', 'jetpack-account-protection' );
+		}
+
+		return '';
+	}
+
+	/**
+	 * Check if the password contains a backslash.
+	 *
+	 * @param string $password The password to check.
+	 *
+	 * @return bool True if the password contains a backslash, false otherwise.
+	 */
+	public function contains_backslash( string $password ): bool {
+		return strpos( $password, '\\' ) !== false;
+	}
+
+	/**
+	 * Check if the password length is within the allowed range.
+	 *
+	 * @param string $password The password to check.
+	 *
+	 * @return bool True if the password is between 6 and 150 characters, false otherwise.
+	 */
+	public function is_invalid_length( string $password ): bool {
+		$length = strlen( $password );
+		return $length < 6 || $length > 150;
+	}
+
+	/**
+	 * Check if the password matches any user data.
+	 *
+	 * @param \WP_User|\stdClass $user The user.
+	 * @param string             $password The password to check.
+	 *
+	 * @return bool True if the password matches any user data, false otherwise.
+	 */
+	public function matches_user_data( $user, string $password ): bool {
```

Comment:
I would recommend making the `$user` parameter required with a single accepted type. Implementation probably shouldn't be calling `matches_user_data` with no user in the first place.

Reply:
Unfortunately, I believe this is necessary and I also wish it weren't. Since we need to use this method both in the `profile_update` (update form and add new user form) action and the `password_reset` action, we end up receiving and needing to reference different data types. Sometimes to check the current changes, to ensure we aren't matching new updates or data inputs in the form for a user that doesn't yet exist and sometimes to reference current user data (reset). For that reason also, we have to pass the `$user` parameter.

Reply:
Ah, I forgot we don't get to use union types until PHP 8 support 😢